src/impl_/concat.rs

```diff
@@ -6,25 +6,210 @@ macro_rules! const_concat {
         $e
     }};
     ($l:expr, $($r:expr),+ $(,)?) => {{
-        const L: &'static str = $l;
-        const R: &'static str = $crate::impl_::concat::const_concat!($($r),*);
-        const LEN: usize = L.len() + R.len();
-        const fn combine(l: &'static [u8], r: &'static [u8]) -> [u8; LEN] {
-            let mut out = [0u8; LEN];
-            let mut i = 0;
-            while i < l.len() {
-                out[i] = l[i];
-                i += 1;
+        const PIECES: &[&str] = &[$l, $($r),+];
+        const RAW_BYTES: &[u8] = &$crate::impl_::concat::combine_to_array::<{
+            $crate::impl_::concat::combined_len(PIECES)
+        }>(PIECES);
+        // Safety: `RAW_BYTES` is combined from valid &str pieces
+        unsafe { ::std::str::from_utf8_unchecked(RAW_BYTES) }
```

Comment:
I rewrote this a bit so that instead of needing to apply the machinery recursively to concatenate pairs, it concatenates a whole slice of pieces.

I also moved as much machinery as possible out into the const fn helpers below. There are some for str inputs and some for bytes inputs (maybe one day with `const AsRef<[u8]>` implementations we could avoid the duplication).

Reply:
Thank you! I think we can migrate the introspection code to use directly these const functions and drop the macro entirely. It might allow us to only use the byte variant of the functions and drop the string variants entirely. Happy to do it in a follow-up.